HSEARCH-3049 Implement the missing work types compared to Search 5 (except DeleteByQuery)

yrodiere opened the pull request:
https://hibernate.atlassian.net//browse/HSEARCH-3049

*Before* you have a look at the commits, I would recommend having a look at the "Explicit indexing" section of the documentation that I just added. Hopefully it will explain why those changes are being made clearly enough :)

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+[NOTE]
+====
+When the the entity mapping to the index changes,
+this rule of thumb can help to determine if the changes require a full reindexing or not:
```
I don't think that those two lines be necessary.

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+When the the entity mapping to the index changes,
+this rule of thumb can help to determine if the changes require a full reindexing or not:
+
+* Wiping the index and reindexing is *not* required if the mapping changes include *only*:
```
I would remove the external list, keeping the internal one.

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+
+As with everything in Hibernate Search,
+these APIs only affect the Hibernate Search indexes:
+they do not write to the database.
```
Maybe /write/write anything/s ?
Or they do not touch the database ?

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+====
+The multi-transaction solution and the periodic `process()` solution can be combined,
+breaking down the process in multiple medium-sized transactions,
+and periodically calling `flush`/`process`/`clear` inside each transaction.
```
Maybe we can add that this last combined strategy is the more flexible and the more suitable for a fine tuning?

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+====
+[source, JAVA, indent=0]
+----
+include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmExplicitIndexingIT.java[tags=write-plan-addOrUpdate]
```
I think there's a wrong tag here.

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+When using multi-tenancy, only documents of one tenant will be removed:
+the tenant of the session from which this writer originated.
+`purgeAsync()`::
+Asynchronous version of `purge()`.
```
I would add `returning a CompletableFuture`.

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+This is generally not useful as Hibernate Search commits changes automatically.
+Only to be used by experts fully aware of the implications.
+`flushAsync()`::
+Asynchronous version of `flush()`.
```
same here

fax4ever commented on documentation/src/main/asciidoc/mapper-orm-indexing-explicit.asciidoc:
```diff
@@ -0,0 +1,374 @@
+If indexes are written to, then calling `optimize()`
+is likely to degrade read/write performance overall.
+`optimizeAsync()`::
+Asynchronous version of `optimize()`.
```
same here

yrodiere commented on mapper/orm/src/main/java/org/hibernate/search/mapper/orm/session/SearchSessionWritePlan.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * An interface for writing to indexes in the context of an ORM Session.
+ * <p>
+ * When a transaction is active, operations are queued internally,
```
At the moment, I don't know. This is a different ticket: https://hibernate.atlassian.net/browse/HSEARCH-3069.

I'll add a TODO here.

yrodiere commented on integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/session/SearchSessionWritePlanIT.java:
```diff
@@ -0,0 +1,390 @@
+			writePlan.purge( IndexedEntity1.class, 6 );
+			writePlan.addOrUpdate( entity6 );
+
+			backend1Mock.expectWorks( IndexedEntity1.INDEX_NAME )
```
I forgot. Done.

yrodiere commented on mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/logging/impl/Log.java:
```diff
@@ -408,4 +413,11 @@ SearchException cannotResolveContainerExtractorName(String extractorName,
+					+ " thus entity with identifier '%2$s' cannot be purged."
+					+ " Use delete() and pass the entity instead of just the identifier."
+	)
+	SearchException cannotPurgeNonIndexedContainedType(
```
We didn't. Done.

yrodiere commented on documentation/src/test/java/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmExplicitIndexingIT.java:
```diff
@@ -0,0 +1,341 @@
+
+	@Test
+	public void persist_automaticIndexing_periodicProcess_multipleTransactions() {
+		EntityManagerFactory entityManagerFactory = setup( HibernateOrmAutomaticIndexingStrategyName.SESSION );
```
Neither one, actually. Fixed.

yrodiere commented:
I addressed most of your comments. Thanks. Waiting for CI and then we can merge.

The only things I did not address were the remarks about the names. I agree they are not great, but there are a few constraints to keep in mind. I will open a ticket so that we can sleep on it and revisit the names before we go Beta.

fax4ever commented:
Merged thanks!
